contracts/LockedNORI.sol

```diff
@@ -40,127 +63,339 @@ contract LockedNORI is ERC777Upgradeable, ERC20PresetMinterPauserUpgradeable, ER
+      address(params.recipient) != address(0),
       "Recipient cannot be zero address"
     );
-    require(super.balanceOf(recipient) == 0, "Recipient has remaining tokens");
-
-    depositFor(recipient, amount);
 
-    TokenGrant storage grant = grants[recipient];
+    TokenGrant storage grant = grants[params.recipient];
     grant.grantAmount = amount;
+    grant.originalAmount = amount;
     grant.exists = true;
-    grant.vestingSchedule.totalAmount = amount;
-    grant.vestingSchedule.startTime = startTime;
-    grant.vestingSchedule.endTime = vestEndTime;
-    // grant.vestingSchedule.addCliff(cliff1Time, vestCliff1Amount);
-    // grant.vestingSchedule.addCliff(cliff2Time, vestCliff2Amount);
+
+    if (params.vestEndTime > params.startTime) {
+      require(params.vestCliff1Amount >= params.unlockCliff1Amount
+        && params.vestCliff2Amount >= params.unlockCliff2Amount,
+        "Unlock cliff amounts cannot exceed vest cliff amounts");
+      grant.vestingSchedule.totalAmount = amount;
+      grant.vestingSchedule.startTime = params.startTime;
+      grant.vestingSchedule.endTime = params.vestEndTime;
+      grant.vestingSchedule.addCliff(params.cliff1Time, params.vestCliff1Amount);
+      grant.vestingSchedule.addCliff(params.cliff2Time, params.vestCliff2Amount);
+    }
 
     grant.lockupSchedule.totalAmount = amount;
-    grant.lockupSchedule.startTime = startTime;
-    grant.lockupSchedule.endTime = vestEndTime;
-    // grant.lockupSchedule.addCliff(cliff1Time, unlockCliff1Amount);
-    // grant.lockupSchedule.addCliff(cliff2Time, unlockCliff2Amount);
+    grant.lockupSchedule.startTime = params.startTime;
+    grant.lockupSchedule.endTime = params.unlockEndTime;
+    grant.lockupSchedule.addCliff(params.cliff1Time, params.unlockCliff1Amount);
+    grant.lockupSchedule.addCliff(params.cliff2Time, params.unlockCliff2Amount);
 
     emit TokenGrantCreated(
-      recipient,
+      params.recipient,
       amount,
-      startTime,
-      vestEndTime,
-      unlockEndTime
+      params.startTime,
+      params.vestEndTime,
+      params.unlockEndTime
     );
   }
 
-  function depositedBalanceOf(address account) public view returns (uint256) {
-    return super.balanceOf(account);
+  /**
+   * @dev grantTo: Sets up an vesting + lockup schedule for recipient.
+   *
+   * Typically this will be invoked via the `tokensReceived` callback for cases
+   * where we have the tokens in hand at the time we set up the grant.
+   *
+   * It is also callable externally to handle the case that tokens are incrementally
+   * deposited after the grant is established.
+   */
+  function grantTo(
+    uint256 amount,
+    address recipient,
+    uint256 startTime,
+    uint256 vestEndTime,
+    uint256 unlockEndTime,
+    uint256 cliff1Time,
+    uint256 cliff2Time,
+    uint256 vestCliff1Amount,
+    uint256 vestCliff2Amount,
+    uint256 unlockCliff1Amount,
+    uint256 unlockCliff2Amount
+  ) public whenNotPaused onlyRole(TOKEN_GRANTER_ROLE) {
```

Comment:
are we currently setting `TOKEN_GRANTER_ROLE` somewhere?

Reply:
thinking this should be `external`

Reply:
along with several other fns

Reply:
We were only assigning that role in the tests. I updated the initialized to give it to the admin as well.  Also switched to external along with others where it made sense.